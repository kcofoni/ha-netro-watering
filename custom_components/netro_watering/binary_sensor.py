"""Support for Netro watering system."""
from __future__ import annotations

from dataclasses import dataclass
import datetime
import logging
from typing import Any

from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
    BinarySensorEntityDescription,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity
import homeassistant.util.dt as dt_util

from .const import (
    CONF_DEVICE_TYPE,
    CONTROLLER_DEVICE_TYPE,
    DOMAIN,
    EXTRA_STATE_ATTRIBUTE_SEP_LEFT,
    EXTRA_STATE_ATTRIBUTE_SEP_RIGHT,
)
from .coordinator import NetroControllerUpdateCoordinator

_LOGGER = logging.getLogger(__name__)


@dataclass
class NetroRequiredKeysMixin:
    """Mixin for required keys."""

    netro_name: str


@dataclass
class NetroBinarySensorEntityDescription(
    BinarySensorEntityDescription, NetroRequiredKeysMixin
):
    """Defines Netro entity description."""


NETRO_ZONE_WATERING_DESCRIPTION = NetroBinarySensorEntityDescription(
    key="iswatering",
    name="Is it watering ?",
    device_class=BinarySensorDeviceClass.RUNNING,
    translation_key="iswatering",
    netro_name="watering",
    icon="mdi:watering-can-outline",
)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up entry for a Netro sensors."""
    if config_entry.data[CONF_DEVICE_TYPE] == CONTROLLER_DEVICE_TYPE:
        controller: NetroControllerUpdateCoordinator = hass.data[DOMAIN][
            config_entry.entry_id
        ]
        # add zone binary sensors
        async_add_entities(
            [
                NetroZone(
                    controller,
                    NETRO_ZONE_WATERING_DESCRIPTION,
                    zone_key,
                )
                for zone_key in controller.active_zones
            ]
        )


class NetroZone(
    CoordinatorEntity[NetroControllerUpdateCoordinator], BinarySensorEntity
):
    """A sensor implementation for Netro Zone device."""

    entity_description: NetroBinarySensorEntityDescription
    _attr_has_entity_name = True

    def __init__(
        self,
        coordinator: NetroControllerUpdateCoordinator,
        description: NetroBinarySensorEntityDescription,
        zone_id: int,
    ) -> None:
        """Initialize the Netro sensor."""
        super().__init__(coordinator)
        self.entity_description = description
        self.zone_id = zone_id
        self._attr_unique_id = (
            f"{coordinator.active_zones[zone_id].serial_number}-{description.key}"
        )
        self._attr_device_info = coordinator.active_zones[zone_id].device_info

    @property
    def is_on(self) -> bool | None:
        """Return True if the current zone is currently watering."""
        return getattr(
            self.coordinator.active_zones[self.zone_id],
            self.entity_description.netro_name,
        )

    @property
    def extra_state_attributes(self) -> dict[str, Any]:
        """Return state attributes."""
        zone_attributes = {
            "zone id": self.zone_id,
            "update interval": f"{round(self.coordinator.update_interval.total_seconds() / 60)} mn"
            if self.coordinator.update_interval is not None
            else None,
        }
        if self.coordinator.current_slowdown_factor > 1:
            zone_attributes[
                "slowdown factor"
            ] = self.coordinator.current_slowdown_factor
        meta_attributes = {
            EXTRA_STATE_ATTRIBUTE_SEP_LEFT: EXTRA_STATE_ATTRIBUTE_SEP_RIGHT,
            "request time (UTC)": self.coordinator.metadata.time
            if self.coordinator.metadata is not None
            else None,
            "last active date": dt_util.as_local(
                self.coordinator.metadata.last_active_date.replace(
                    tzinfo=datetime.UTC
                )
            )
            if self.coordinator.metadata is not None
            else None,
            "transaction id": self.coordinator.metadata.tid
            if self.coordinator.metadata is not None
            else None,
            "token limit": self.coordinator.metadata.token_limit
            if self.coordinator.metadata is not None
            else None,
            "token remaining": self.coordinator.metadata.token_remaining
            if self.coordinator.metadata is not None
            else None,
            "token reset": dt_util.as_local(
                self.coordinator.metadata.token_reset_date.replace(
                    tzinfo=datetime.UTC
                )
            )
            if self.coordinator.metadata is not None
            else None,
        }
        return zone_attributes | meta_attributes
