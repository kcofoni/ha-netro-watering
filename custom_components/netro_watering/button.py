"""Support du bouton Refresh pour Netro Watering."""

from __future__ import annotations

import logging

from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import DOMAIN
from .coordinator import NetroControllerUpdateCoordinator

_LOGGER = logging.getLogger(__name__)

NETRO_CONTROLLER_BUTTON_DESCRIPTION = ButtonEntityDescription(
    key="refresh",
    name="Refresh",
    entity_registry_enabled_default=True,
    translation_key="refresh",
    icon="mdi:refresh",
)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
) -> None:
    """Set up the button platform."""
    controller: NetroControllerUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
    _LOGGER.info("Adding button entity for controller = %s", controller.device_name)

    async_add_entities(
        [NetroRefreshButton(controller, NETRO_CONTROLLER_BUTTON_DESCRIPTION)]
    )


class NetroRefreshButton(
    CoordinatorEntity[NetroControllerUpdateCoordinator], ButtonEntity
):
    """Bouton qui force un refresh du contrôleur Netro."""

    entity_description: ButtonEntityDescription
    _attr_has_entity_name = True

    def __init__(
        self,
        coordinator: NetroControllerUpdateCoordinator,
        description: ButtonEntityDescription,
    ) -> None:
        """Initialize the Netro refresh button entity.

        Args:
            coordinator: The data update coordinator for Netro.
            description: The entity description for the button.
        """
        super().__init__(coordinator)
        self.entity_description = description
        self._attr_unique_id = f"{coordinator.serial_number}-{description.key}"
        self._attr_device_info = coordinator.device_info

    async def async_press(self) -> None:
        """Pressing the button requests an immediate update."""
        _LOGGER.debug(
            "Netro: refresh button pressed for %s", self.coordinator.serial_number
        )

        # Si ton coordinator expose une méthode spécifique (ex: refresh_controller()),
        # tu peux l'appeler ici avant/à la place du request_refresh.
        # await self.coordinator.refresh_controller()

        # Dans tous les cas, on (re)planifie une mise à jour via le DataUpdateCoordinator
        await self.coordinator.async_request_refresh()
