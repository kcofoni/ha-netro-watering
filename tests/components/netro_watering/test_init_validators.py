"""Tests for validator functions in __init__.py."""

import pytest
import voluptuous as vol

from custom_components.netro_watering import _validate_slowdown_factors, hhmm_or_hhmmss


class TestValidateSlowdownFactors:
    """Test suite for _validate_slowdown_factors function."""

    def test_validate_slowdown_factors_none_input(self, snapshot):
        """Test _validate_slowdown_factors with None input."""
        result = _validate_slowdown_factors(None)

        assert result == snapshot

    def test_validate_slowdown_factors_empty_list(self, snapshot):
        """Test _validate_slowdown_factors with empty list."""
        result = _validate_slowdown_factors([])

        assert result == snapshot

    def test_validate_slowdown_factors_invalid_type(self, snapshot):
        """Test _validate_slowdown_factors with non-list input."""
        with pytest.raises(vol.Invalid) as exc_info:
            _validate_slowdown_factors("not a list")

        result = {
            "error_message": str(exc_info.value),
        }

        assert result == snapshot

    def test_validate_slowdown_factors_valid_basic_case(self, snapshot):
        """Test _validate_slowdown_factors with valid basic slowdown factors."""
        input_data = [
            {"from": "08:00", "to": "12:00", "sdf": 2},
            {"from": "14:00", "to": "18:00", "sdf": 3},
        ]

        result = _validate_slowdown_factors(input_data)

        # Verify original input is returned unchanged
        assert result is input_data
        assert result == snapshot

    def test_validate_slowdown_factors_valid_with_seconds(self, snapshot):
        """Test _validate_slowdown_factors with HH:MM:SS format."""
        input_data = [
            {"from": "08:00:30", "to": "12:30:45", "sdf": 2},
            {"from": "20:15:00", "to": "23:59:59", "sdf": 4},
        ]

        result = _validate_slowdown_factors(input_data)

        assert result is input_data
        assert result == snapshot

    def test_validate_slowdown_factors_midnight_crossing(self, snapshot):
        """Test _validate_slowdown_factors with midnight crossing (allowed)."""
        input_data = [
            {"from": "22:00", "to": "06:00", "sdf": 2},  # Crosses midnight
            {"from": "23:30", "to": "01:30", "sdf": 3},  # Crosses midnight
        ]

        result = _validate_slowdown_factors(input_data)

        assert result is input_data
        assert result == snapshot

    def test_validate_slowdown_factors_character_cleaning(self, snapshot):
        """Test _validate_slowdown_factors cleans NBSP and ZWSP characters."""
        input_data = [
            {
                "from": "08:00\u00a0",  # NBSP at end
                "to": "\u200b12:00",  # ZWSP at start
                "sdf": 2,
            },
            {
                "from": " 14:00 ",  # Regular spaces
                "to": "18:00\u00a0\u200b",  # Both special chars
                "sdf": 3,
            },
        ]

        result = _validate_slowdown_factors(input_data)

        # Original input should be preserved despite cleaning for validation
        assert result is input_data
        assert result == snapshot

    def test_validate_slowdown_factors_invalid_time_format(self, snapshot):
        """Test _validate_slowdown_factors with invalid time formats."""
        input_data = [
            {"from": "25:00", "to": "12:00", "sdf": 2},  # Invalid hour
        ]

        with pytest.raises(vol.Invalid) as exc_info:
            _validate_slowdown_factors(input_data)

        result = {
            "error_message": str(exc_info.value),
        }

        assert result == snapshot

    def test_validate_slowdown_factors_invalid_time_format_minutes(self, snapshot):
        """Test _validate_slowdown_factors with invalid minutes."""
        input_data = [
            {"from": "08:60", "to": "12:00", "sdf": 2},  # Invalid minutes
        ]

        with pytest.raises(vol.Invalid) as exc_info:
            _validate_slowdown_factors(input_data)

        result = {
            "error_message": str(exc_info.value),
        }

        assert result == snapshot

    def test_validate_slowdown_factors_malformed_time_string(self, snapshot):
        """Test _validate_slowdown_factors with malformed time strings."""
        input_data = [
            {
                "from": "8:0",
                "to": "12:00",
                "sdf": 2,
            },  # Actually valid - Python accepts it
        ]

        # This should work since Python's strptime is flexible
        result = _validate_slowdown_factors(input_data)

        assert result is input_data
        assert result == snapshot

    def test_validate_slowdown_factors_non_time_string(self, snapshot):
        """Test _validate_slowdown_factors with completely invalid time strings."""
        input_data = [
            {"from": "not a time", "to": "12:00", "sdf": 2},
        ]

        with pytest.raises(vol.Invalid) as exc_info:
            _validate_slowdown_factors(input_data)

        result = {
            "error_message": str(exc_info.value),
        }

        assert result == snapshot

    def test_validate_slowdown_factors_sdf_less_than_one(self, snapshot):
        """Test _validate_slowdown_factors with sdf < 1."""
        input_data = [
            {"from": "08:00", "to": "12:00", "sdf": 0},  # sdf = 0 invalid
        ]

        with pytest.raises(vol.Invalid) as exc_info:
            _validate_slowdown_factors(input_data)

        result = {
            "error_message": str(exc_info.value),
        }

        assert result == snapshot

    def test_validate_slowdown_factors_negative_sdf(self, snapshot):
        """Test _validate_slowdown_factors with negative sdf."""
        input_data = [
            {"from": "08:00", "to": "12:00", "sdf": -1},  # Negative sdf
        ]

        with pytest.raises(vol.Invalid) as exc_info:
            _validate_slowdown_factors(input_data)

        result = {
            "error_message": str(exc_info.value),
        }

        assert result == snapshot

    def test_validate_slowdown_factors_identical_times(self, snapshot):
        """Test _validate_slowdown_factors with identical from and to times."""
        input_data = [
            {"from": "12:00", "to": "12:00", "sdf": 2},  # Identical times forbidden
        ]

        with pytest.raises(vol.Invalid) as exc_info:
            _validate_slowdown_factors(input_data)

        result = {
            "error_message": str(exc_info.value),
        }

        assert result == snapshot

    def test_validate_slowdown_factors_identical_times_with_seconds(self, snapshot):
        """Test _validate_slowdown_factors with identical times including seconds."""
        input_data = [
            {"from": "12:30:45", "to": "12:30:45", "sdf": 2},  # Identical with seconds
        ]

        with pytest.raises(vol.Invalid) as exc_info:
            _validate_slowdown_factors(input_data)

        result = {
            "error_message": str(exc_info.value),
        }

        assert result == snapshot

    def test_validate_slowdown_factors_multiple_items_valid(self, snapshot):
        """Test _validate_slowdown_factors with multiple valid items."""
        input_data = [
            {"from": "06:00", "to": "09:00", "sdf": 1},
            {"from": "09:30", "to": "12:30", "sdf": 2},
            {"from": "14:00:00", "to": "17:30:30", "sdf": 3},
            {"from": "20:00", "to": "23:59", "sdf": 4},
            {"from": "23:30", "to": "02:00", "sdf": 2},  # Crosses midnight
        ]

        result = _validate_slowdown_factors(input_data)

        assert result is input_data
        assert result == snapshot

    def test_validate_slowdown_factors_error_includes_index(self, snapshot):
        """Test _validate_slowdown_factors error messages include correct index."""
        input_data = [
            {"from": "08:00", "to": "12:00", "sdf": 2},  # Valid
            {"from": "14:00", "to": "18:00", "sdf": 3},  # Valid
            {
                "from": "20:00",
                "to": "20:00",
                "sdf": 2,
            },  # Invalid (identical times) - index 2
        ]

        with pytest.raises(vol.Invalid) as exc_info:
            _validate_slowdown_factors(input_data)

        result = {
            "error_message": str(exc_info.value),
            "contains_index_2": "[2]" in str(exc_info.value),
        }

        assert result == snapshot

    def test_validate_slowdown_factors_mixed_time_formats(self, snapshot):
        """Test _validate_slowdown_factors with mixed HH:MM and HH:MM:SS formats."""
        input_data = [
            {"from": "08:00", "to": "12:30:45", "sdf": 2},  # Mixed formats
            {"from": "14:15:30", "to": "18:00", "sdf": 3},  # Mixed formats
            {"from": "20:00:00", "to": "23:59:59", "sdf": 1},  # Both with seconds
        ]

        result = _validate_slowdown_factors(input_data)

        assert result is input_data
        assert result == snapshot

    def test_validate_slowdown_factors_boundary_times(self, snapshot):
        """Test _validate_slowdown_factors with boundary time values."""
        input_data = [
            {"from": "00:00", "to": "23:59", "sdf": 1},  # Full day range
            {"from": "00:00:00", "to": "23:59:59", "sdf": 2},  # Full day with seconds
            {"from": "23:59", "to": "00:01", "sdf": 3},  # Minute crossing midnight
        ]

        result = _validate_slowdown_factors(input_data)

        assert result is input_data
        assert result == snapshot

    def test_validate_slowdown_factors_sdf_edge_values(self, snapshot):
        """Test _validate_slowdown_factors with edge sdf values."""
        input_data = [
            {"from": "08:00", "to": "12:00", "sdf": 1},  # Minimum valid sdf
            {"from": "14:00", "to": "18:00", "sdf": 100},  # Large valid sdf
        ]

        result = _validate_slowdown_factors(input_data)

        assert result is input_data
        assert result == snapshot


class TestHhmm_or_hhmmss:
    """Test suite for hhmm_or_hhmmss function."""

    def test_hhmm_or_hhmmss_valid_hhmm_format(self, snapshot):
        """Test hhmm_or_hhmmss with valid HH:MM format."""
        test_cases = [
            "00:00",
            "08:30",
            "12:00",
            "23:59",
            "09:15",
        ]

        results = []
        for test_time in test_cases:
            result = hhmm_or_hhmmss(test_time)
            # Verify original string is returned unchanged
            assert result is test_time
            results.append(result)

        assert results == snapshot

    def test_hhmm_or_hhmmss_valid_hhmmss_format(self, snapshot):
        """Test hhmm_or_hhmmss with valid HH:MM:SS format."""
        test_cases = [
            "00:00:00",
            "08:30:45",
            "12:00:30",
            "23:59:59",
            "09:15:22",
        ]

        results = []
        for test_time in test_cases:
            result = hhmm_or_hhmmss(test_time)
            # Verify original string is returned unchanged
            assert result is test_time
            results.append(result)

        assert results == snapshot

    def test_hhmm_or_hhmmss_with_special_characters(self, snapshot):
        """Test hhmm_or_hhmmss with special characters that get cleaned."""
        test_cases = [
            " 08:30 ",  # Leading/trailing spaces
            "\u00a008:30\u00a0",  # NBSP characters
            "\u200b08:30\u200b",  # ZWSP characters
            " \u00a0\u200b08:30\u200b\u00a0 ",  # Mixed special chars
        ]

        results = []
        for test_time in test_cases:
            result = hhmm_or_hhmmss(test_time)
            # Verify original string is returned unchanged (not cleaned)
            assert result is test_time
            results.append(result)

        assert results == snapshot

    def test_hhmm_or_hhmmss_invalid_non_string_types(self, snapshot):
        """Test hhmm_or_hhmmss with non-string input types."""
        test_cases = [
            123,
            12.34,
            None,
            [],
            {},
            True,
        ]

        results = []
        for test_input in test_cases:
            with pytest.raises(vol.Invalid) as exc_info:
                hhmm_or_hhmmss(test_input)
            results.append(
                {
                    "input": str(test_input),
                    "input_type": type(test_input).__name__,
                    "error_message": str(exc_info.value),
                }
            )

        assert results == snapshot

    def test_hhmm_or_hhmmss_invalid_time_formats(self, snapshot):
        """Test hhmm_or_hhmmss with invalid time format strings."""
        test_cases = [
            "24:00",  # Invalid hour (24)
            "12:60",  # Invalid minute (60)
            "12:30:60",  # Invalid second (60)
            "8:30",  # Missing leading zero for hour
            "08:5",  # Missing leading zero for minute
            "12:30:5",  # Missing leading zero for second
            "abc:def",  # Non-numeric
            "12:30:45:00",  # Too many parts
            "12",  # Missing minute
            ":30",  # Missing hour
            "12:",  # Missing minute after colon
            "12:30:",  # Missing second after colon
            "",  # Empty string
            "25:30",  # Hour > 23
            "12:70",  # Minute > 59
            "12:30:70",  # Second > 59
            "-1:30",  # Negative hour
            "12:-5",  # Negative minute
            "12:30:-1",  # Negative second
        ]

        results = []
        for test_time in test_cases:
            with pytest.raises(vol.Invalid) as exc_info:
                hhmm_or_hhmmss(test_time)
            results.append(
                {
                    "input": test_time,
                    "error_message": str(exc_info.value),
                }
            )

        assert results == snapshot

    def test_hhmm_or_hhmmss_edge_cases(self, snapshot):
        """Test hhmm_or_hhmmss with edge cases."""
        # Valid edge cases
        valid_cases = [
            "00:00",  # Midnight
            "23:59",  # One minute before midnight
            "00:00:00",  # Midnight with seconds
            "23:59:59",  # Last second of day
        ]

        valid_results = []
        for test_time in valid_cases:
            result = hhmm_or_hhmmss(test_time)
            assert result is test_time
            valid_results.append(result)

        # Invalid edge cases
        invalid_cases = [
            "24:00",  # Invalid midnight representation
            "23:60",  # Invalid minute
            "24:00:00",  # Invalid midnight with seconds
            "23:59:60",  # Invalid second
        ]

        invalid_results = []
        for test_time in invalid_cases:
            with pytest.raises(vol.Invalid) as exc_info:
                hhmm_or_hhmmss(test_time)
            invalid_results.append(
                {
                    "input": test_time,
                    "error_message": str(exc_info.value),
                }
            )

        result = {
            "valid_cases": valid_results,
            "invalid_cases": invalid_results,
        }

        assert result == snapshot

    def test_hhmm_or_hhmmss_preserves_original_string(self, snapshot):
        """Test that hhmm_or_hhmmss always returns the exact original string."""
        # Test cases with various formatting that should be cleaned but original preserved
        test_cases = [
            "08:30",  # Clean format
            " 08:30 ",  # With spaces
            "\u00a008:30\u00a0",  # With NBSP
            "\u200b08:30\u200b",  # With ZWSP
            " \u00a008:30\u00a0 ",  # Mixed whitespace
        ]

        results = []
        for original in test_cases:
            result = hhmm_or_hhmmss(original)
            # Critical test: result must be the EXACT same object
            assert result is original
            # And also equal in value
            assert result == original
            results.append(
                {
                    "original": repr(original),
                    "result": repr(result),
                    "is_same_object": result is original,
                    "is_equal_value": result == original,
                }
            )

        assert results == snapshot
