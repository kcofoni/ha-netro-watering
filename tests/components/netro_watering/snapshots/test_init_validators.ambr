# serializer version: 1
# name: TestHhmm_or_hhmmss.test_hhmm_or_hhmmss_edge_cases
  dict({
    'invalid_cases': list([
      dict({
        'error_message': 'Expected time in HH:MM or HH:MM:SS',
        'input': '24:00',
      }),
      dict({
        'error_message': 'Expected time in HH:MM or HH:MM:SS',
        'input': '23:60',
      }),
      dict({
        'error_message': 'Expected time in HH:MM or HH:MM:SS',
        'input': '24:00:00',
      }),
      dict({
        'error_message': 'Expected time in HH:MM or HH:MM:SS',
        'input': '23:59:60',
      }),
    ]),
    'valid_cases': list([
      '00:00',
      '23:59',
      '00:00:00',
      '23:59:59',
    ]),
  })
# ---
# name: TestHhmm_or_hhmmss.test_hhmm_or_hhmmss_invalid_non_string_types
  list([
    dict({
      'error_message': 'Expected time as string in HH:MM or HH:MM:SS',
      'input': '123',
      'input_type': 'int',
    }),
    dict({
      'error_message': 'Expected time as string in HH:MM or HH:MM:SS',
      'input': '12.34',
      'input_type': 'float',
    }),
    dict({
      'error_message': 'Expected time as string in HH:MM or HH:MM:SS',
      'input': 'None',
      'input_type': 'NoneType',
    }),
    dict({
      'error_message': 'Expected time as string in HH:MM or HH:MM:SS',
      'input': '[]',
      'input_type': 'list',
    }),
    dict({
      'error_message': 'Expected time as string in HH:MM or HH:MM:SS',
      'input': '{}',
      'input_type': 'dict',
    }),
    dict({
      'error_message': 'Expected time as string in HH:MM or HH:MM:SS',
      'input': 'True',
      'input_type': 'bool',
    }),
  ])
# ---
# name: TestHhmm_or_hhmmss.test_hhmm_or_hhmmss_invalid_time_formats
  list([
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '24:00',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:60',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:30:60',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '8:30',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '08:5',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:30:5',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': 'abc:def',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:30:45:00',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': ':30',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:30:',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '25:30',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:70',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:30:70',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '-1:30',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:-5',
    }),
    dict({
      'error_message': 'Expected time in HH:MM or HH:MM:SS',
      'input': '12:30:-1',
    }),
  ])
# ---
# name: TestHhmm_or_hhmmss.test_hhmm_or_hhmmss_preserves_original_string
  list([
    dict({
      'is_equal_value': True,
      'is_same_object': True,
      'original': "'08:30'",
      'result': "'08:30'",
    }),
    dict({
      'is_equal_value': True,
      'is_same_object': True,
      'original': "' 08:30 '",
      'result': "' 08:30 '",
    }),
    dict({
      'is_equal_value': True,
      'is_same_object': True,
      'original': "'\\xa008:30\\xa0'",
      'result': "'\\xa008:30\\xa0'",
    }),
    dict({
      'is_equal_value': True,
      'is_same_object': True,
      'original': "'\\u200b08:30\\u200b'",
      'result': "'\\u200b08:30\\u200b'",
    }),
    dict({
      'is_equal_value': True,
      'is_same_object': True,
      'original': "' \\xa008:30\\xa0 '",
      'result': "' \\xa008:30\\xa0 '",
    }),
  ])
# ---
# name: TestHhmm_or_hhmmss.test_hhmm_or_hhmmss_valid_hhmm_format
  list([
    '00:00',
    '08:30',
    '12:00',
    '23:59',
    '09:15',
  ])
# ---
# name: TestHhmm_or_hhmmss.test_hhmm_or_hhmmss_valid_hhmmss_format
  list([
    '00:00:00',
    '08:30:45',
    '12:00:30',
    '23:59:59',
    '09:15:22',
  ])
# ---
# name: TestHhmm_or_hhmmss.test_hhmm_or_hhmmss_with_special_characters
  list([
    ' 08:30 ',
    '\xa008:30\xa0',
    '\u200b08:30\u200b',
    ' \xa0\u200b08:30\u200b\xa0 ',
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_boundary_times
  list([
    dict({
      'from': '00:00',
      'sdf': 1,
      'to': '23:59',
    }),
    dict({
      'from': '00:00:00',
      'sdf': 2,
      'to': '23:59:59',
    }),
    dict({
      'from': '23:59',
      'sdf': 3,
      'to': '00:01',
    }),
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_character_cleaning
  list([
    dict({
      'from': '08:00\xa0',
      'sdf': 2,
      'to': '\u200b12:00',
    }),
    dict({
      'from': ' 14:00 ',
      'sdf': 3,
      'to': '18:00\xa0\u200b',
    }),
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_empty_list
  list([
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_error_includes_index
  dict({
    'contains_index_2': True,
    'error_message': "slowdown_factors[2]: 'from' and 'to' must not be identical",
  })
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_identical_times
  dict({
    'error_message': "slowdown_factors[0]: 'from' and 'to' must not be identical",
  })
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_identical_times_with_seconds
  dict({
    'error_message': "slowdown_factors[0]: 'from' and 'to' must not be identical",
  })
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_invalid_time_format
  dict({
    'error_message': "Expected time in HH:MM or HH:MM:SS (e.g., '23:05' or '23:05:30')",
  })
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_invalid_time_format_minutes
  dict({
    'error_message': "Expected time in HH:MM or HH:MM:SS (e.g., '23:05' or '23:05:30')",
  })
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_invalid_type
  dict({
    'error_message': 'slowdown_factors must be a list',
  })
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_malformed_time_string
  list([
    dict({
      'from': '8:0',
      'sdf': 2,
      'to': '12:00',
    }),
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_midnight_crossing
  list([
    dict({
      'from': '22:00',
      'sdf': 2,
      'to': '06:00',
    }),
    dict({
      'from': '23:30',
      'sdf': 3,
      'to': '01:30',
    }),
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_mixed_time_formats
  list([
    dict({
      'from': '08:00',
      'sdf': 2,
      'to': '12:30:45',
    }),
    dict({
      'from': '14:15:30',
      'sdf': 3,
      'to': '18:00',
    }),
    dict({
      'from': '20:00:00',
      'sdf': 1,
      'to': '23:59:59',
    }),
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_multiple_items_valid
  list([
    dict({
      'from': '06:00',
      'sdf': 1,
      'to': '09:00',
    }),
    dict({
      'from': '09:30',
      'sdf': 2,
      'to': '12:30',
    }),
    dict({
      'from': '14:00:00',
      'sdf': 3,
      'to': '17:30:30',
    }),
    dict({
      'from': '20:00',
      'sdf': 4,
      'to': '23:59',
    }),
    dict({
      'from': '23:30',
      'sdf': 2,
      'to': '02:00',
    }),
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_negative_sdf
  dict({
    'error_message': 'slowdown_factors[0].sdf must be >= 1',
  })
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_non_time_string
  dict({
    'error_message': "Expected time in HH:MM or HH:MM:SS (e.g., '23:05' or '23:05:30')",
  })
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_none_input
  list([
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_sdf_edge_values
  list([
    dict({
      'from': '08:00',
      'sdf': 1,
      'to': '12:00',
    }),
    dict({
      'from': '14:00',
      'sdf': 100,
      'to': '18:00',
    }),
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_sdf_less_than_one
  dict({
    'error_message': 'slowdown_factors[0].sdf must be >= 1',
  })
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_valid_basic_case
  list([
    dict({
      'from': '08:00',
      'sdf': 2,
      'to': '12:00',
    }),
    dict({
      'from': '14:00',
      'sdf': 3,
      'to': '18:00',
    }),
  ])
# ---
# name: TestValidateSlowdownFactors.test_validate_slowdown_factors_valid_with_seconds
  list([
    dict({
      'from': '08:00:30',
      'sdf': 2,
      'to': '12:30:45',
    }),
    dict({
      'from': '20:15:00',
      'sdf': 4,
      'to': '23:59:59',
    }),
  ])
# ---
